libraries/AP_IOMCU/iofirmware/analog.cpp

```diff
@@ -26,63 +26,105 @@
 #error "HAL_USE_ADC must be set"
 #endif
 
+extern "C" {
+    extern void Vector88();
+}
+
+#define STM32_ADC1_NUMBER 18
+#define STM32_ADC1_HANDLER Vector88
+
+const uint32_t VSERVO_CHANNEL = ADC_SQR3_SQ1_N(ADC_CHANNEL_IN4);
+const uint32_t VRSSI_CHANNEL = ADC_SQR3_SQ1_N(ADC_CHANNEL_IN5);
+
+static uint16_t vrssi_val = 0xFFFF;
+static uint16_t vservo_val = 0xFFFF;
+
 /*
   initialise ADC capture
  */
 void adc_init(void)
 {
-    adc_lld_init();
     rccEnableADC1(true);
+    ADC1->CR1 = 0;
+    ADC1->CR2 = ADC_CR2_ADON;
+
+    /* Reset calibration just to be safe.*/
+    ADC1->CR2 = ADC_CR2_ADON | ADC_CR2_RSTCAL;
+    while ((ADC1->CR2 & ADC_CR2_RSTCAL) != 0)
+      ;
+
+    /* Calibration.*/
+    ADC1->CR2 = ADC_CR2_ADON | ADC_CR2_CAL;
+    while ((ADC1->CR2 & ADC_CR2_CAL) != 0)
+      ;
 
     /* set channels 4 and 5 for 28.5us sample time */
     ADC1->SMPR2 = ADC_SMPR2_SMP_AN4(ADC_SAMPLE_28P5) | ADC_SMPR2_SMP_AN5(ADC_SAMPLE_28P5);
 
     /* capture a single sample at a time */
     ADC1->SQR1 = 0;
     ADC1->SQR2 = 0;
+
+    ADC1->CR1 |= ADC_CR1_EOCIE;
+
+    nvicEnableVector(STM32_ADC1_NUMBER, STM32_ADC_ADC1_IRQ_PRIORITY);
 }
 
 /*
   capture one sample on a channel
  */
-static uint16_t adc_sample_channel(uint32_t channel)
+static void adc_sample_channel(uint32_t channel)
 {
-    // clear EOC
-    ADC1->SR = 0;
+    chSysLock();
 
-    /* capture one sample */
-    ADC1->SQR3 = channel;
-    ADC1->CR2 |= ADC_CR2_ADON;
-
-    /* wait for the conversion to complete */
-    uint32_t counter = 16;
-
-    while (!(ADC1->SR & ADC_SR_EOC)) {
-        if (--counter == 0) {
-            // ensure EOC is clear
-            ADC1->SR = 0;
-            return 0xffff;
-        }
+    if (ADC1->SR & ADC_SR_STRT) {
```

Comment:
I think this will happen a lot, and may cause us to get unreliable vservo/rssi updates. We call adc_sample_vrssi() and adc_sample_vservo() in quick succession, so the ADC will be busy most of the time on the vservo update if rssi sampling is done.
I think we need an adc_sample_next() call or similar which toggles between the two.

Reply:
yeah good call, the API seemed a bit anomalous with the interrupt changes

Reply:
Fixed